File: backend/app/schemas/phase4.py
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, field_serializer, model_validator

from app.models.manufacturing import (
    InventoryKind,
    InventoryTransactionType,
    ManufacturingStatus,
    ShipmentStatus,
)
from app.schemas.manufacturing import Quantity


class ShipmentLineInput(BaseModel):
    product_id: int = Field(gt=0)
    quantity: Quantity


class ShipmentWrite(BaseModel):
    shipment_number: str = Field(min_length=1, max_length=30)
    shipped_date: date
    lines: list[ShipmentLineInput] = Field(min_length=1)

    @model_validator(mode="after")
    def products_are_unique(self):
        product_ids = [line.product_id for line in self.lines]
        if len(product_ids) != len(set(product_ids)):
            raise ValueError("同じ製品を重複指定できません。")
        return self


class ShipmentLineResponse(ShipmentLineInput):
    model_config = ConfigDict(from_attributes=True)
    id: int
    product_code: str
    product_name: str

    @field_serializer("quantity", when_used="json")
    def serialize_quantity(self, value: Decimal) -> float:
        return float(value)


class ShipmentResponse(BaseModel):
    id: int
    shipment_number: str
    shipped_date: date
    status: ShipmentStatus
    confirmed_at: datetime | None
    created_at: datetime
    updated_at: datetime
    lines: list[ShipmentLineResponse]


class ShipmentListResponse(BaseModel):
    items: list[ShipmentResponse]
    page: int
    page_size: int
    total: int
    total_pages: int


class RawBalanceResponse(BaseModel):
    id: int
    tea_leaf_id: int
    tea_leaf_code: str
    tea_leaf_name: str
    variety_id: int
    variety_code: str
    variety_name: str
    quantity: Decimal
    unit: str = "kg"
    updated_at: datetime

    @field_serializer("quantity", when_used="json")
    def serialize_quantity(self, value: Decimal) -> float:
        return float(value)


class ProductBalanceResponse(BaseModel):
    id: int
    product_id: int
    product_code: str
    product_name: str
    quantity: Decimal
    unit: str = "kg"
    updated_at: datetime

    @field_serializer("quantity", when_used="json")
    def serialize_quantity(self, value: Decimal) -> float:
        return float(value)


class InventoryTransactionResponse(BaseModel):
    id: int
    inventory_kind: InventoryKind
    transaction_type: InventoryTransactionType
    tea_leaf_id: int | None
    variety_id: int | None
    product_id: int | None
    target_code: str
    target_name: str
    quantity_delta: Decimal
    balance_after: Decimal
    unit: str = "kg"
    reference_type: str
    reference_id: int
    occurred_at: datetime

    @field_serializer("quantity_delta", "balance_after", when_used="json")
    def serialize_quantity(self, value: Decimal) -> float:
        return float(value)


class PageResponse(BaseModel):
    items: list
    page: int
    page_size: int
    total: int
    total_pages: int


class BreakdownItem(BaseModel):
    code: str
    name: str
    quantity: Decimal

    @field_serializer("quantity", when_used="json")
    def serialize_quantity(self, value: Decimal) -> float:
        return float(value)


class SummaryResponse(BaseModel):
    date_from: date
    date_to: date
    receipt_quantity: Decimal
    manufacturing_quantity: Decimal
    shipment_quantity: Decimal
    current_raw_material_quantity: Decimal
    current_product_quantity: Decimal
    receipt_breakdown: list[BreakdownItem]
    manufacturing_breakdown: list[BreakdownItem]
    shipment_breakdown: list[BreakdownItem]

    @field_serializer(
        "receipt_quantity",
        "manufacturing_quantity",
        "shipment_quantity",
        "current_raw_material_quantity",
        "current_product_quantity",
        when_used="json",
    )
    def serialize_quantity(self, value: Decimal) -> float:
        return float(value)


class InventoryOverview(BaseModel):
    item_count: int
    total_quantity: Decimal
    unit: str = "kg"

    @field_serializer("total_quantity", when_used="json")
    def serialize_quantity(self, value: Decimal) -> float:
        return float(value)


class DashboardResponse(BaseModel):
    date_from: date
    date_to: date
    manufacturing_status_counts: dict[ManufacturingStatus, int]
    raw_material_inventory: InventoryOverview
    product_inventory: InventoryOverview
    receipt_quantity: Decimal
    manufacturing_quantity: Decimal
    shipment_quantity: Decimal

    @field_serializer(
        "receipt_quantity", "manufacturing_quantity", "shipment_quantity", when_used="json"
    )
    def serialize_quantity(self, value: Decimal) -> float:
        return float(value)

File: frontend/src/api/phase4.ts
import { apiFetch } from "./client";

export type Page<T> = {
  items: T[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
};

export type RawBalance = {
  id: number;
  tea_leaf_id: number;
  tea_leaf_code: string;
  tea_leaf_name: string;
  variety_id: number;
  variety_code: string;
  variety_name: string;
  quantity: number;
  unit: "kg";
  updated_at: string;
};

export type ProductBalance = {
  id: number;
  product_id: number;
  product_code: string;
  product_name: string;
  quantity: number;
  unit: "kg";
  updated_at: string;
};

export type InventoryTransaction = {
  id: number;
  inventory_kind: "RAW_MATERIAL" | "PRODUCT";
  transaction_type:
    | "RECEIPT"
    | "MANUFACTURING_CONSUMPTION"
    | "MANUFACTURING_OUTPUT"
    | "SHIPMENT";
  target_code: string;
  target_name: string;
  quantity_delta: number;
  balance_after: number;
  unit: "kg";
  occurred_at: string;
};

export type ShipmentInput = {
  shipment_number: string;
  shipped_date: string;
  lines: Array<{ product_id: number; quantity: string }>;
};

export type Shipment = {
  id: number;
  shipment_number: string;
  shipped_date: string;
  status: "DRAFT" | "CONFIRMED";
  confirmed_at: string | null;
  created_at: string;
  updated_at: string;
  lines: Array<{
    id: number;
    product_id: number;
    product_code: string;
    product_name: string;
    quantity: number;
  }>;
};

export type Breakdown = { code: string; name: string; quantity: number };
export type Summary = {
  date_from: string;
  date_to: string;
  receipt_quantity: number;
  manufacturing_quantity: number;
  shipment_quantity: number;
  current_raw_material_quantity: number;
  current_product_quantity: number;
  receipt_breakdown: Breakdown[];
  manufacturing_breakdown: Breakdown[];
  shipment_breakdown: Breakdown[];
};
export type Dashboard = {
  date_from: string;
  date_to: string;
  manufacturing_status_counts: Record<string, number>;
  raw_material_inventory: {
    item_count: number;
    total_quantity: number;
    unit: "kg";
  };
  product_inventory: { item_count: number; total_quantity: number; unit: "kg" };
  receipt_quantity: number;
  manufacturing_quantity: number;
  shipment_quantity: number;
};

const json = (method: "POST" | "PUT", body?: unknown): RequestInit => ({
  method,
  headers:
    body === undefined ? undefined : { "Content-Type": "application/json" },
  body: body === undefined ? undefined : JSON.stringify(body),
});

export const fetchRawBalances = (page = 1, pageSize = 20) =>
  apiFetch<Page<RawBalance>>(
    `/inventories/raw-materials?page=${page}&page_size=${pageSize}`,
  );
export const fetchProductBalances = (page = 1, pageSize = 20) =>
  apiFetch<Page<ProductBalance>>(
    `/inventories/products?page=${page}&page_size=${pageSize}`,
  );
export const fetchInventoryTransactions = (query: string) =>
  apiFetch<Page<InventoryTransaction>>(`/inventory-transactions?${query}`);
export const fetchShipments = (page = 1, pageSize = 20) =>
  apiFetch<Page<Shipment>>(`/shipments?page=${page}&page_size=${pageSize}`);
export const fetchShipment = (id: string) =>
  apiFetch<Shipment>(`/shipments/${id}`);
export const createShipment = (body: ShipmentInput) =>
  apiFetch<Shipment>("/shipments", json("POST", body));
export const updateShipment = (id: number, body: ShipmentInput) =>
  apiFetch<Shipment>(`/shipments/${id}`, json("PUT", body));
export const confirmShipment = (id: number) =>
  apiFetch<Shipment>(`/shipments/${id}/confirm`, json("POST"));
export const fetchSummary = (dateFrom: string, dateTo: string) =>
  apiFetch<Summary>(`/reports/summary?date_from=${dateFrom}&date_to=${dateTo}`);
export const fetchDashboard = (dateFrom: string, dateTo: string) =>
  apiFetch<Dashboard>(`/dashboard?date_from=${dateFrom}&date_to=${dateTo}`);

export function tokyoToday(): string {
  return new Intl.DateTimeFormat("sv-SE", {
    timeZone: "Asia/Tokyo",
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  }).format(new Date());
}

export function defaultPeriod(): { dateFrom: string; dateTo: string } {
  const dateTo = tokyoToday();
  const start = new Date(`${dateTo}T00:00:00+09:00`);
  start.setUTCDate(start.getUTCDate() - 29);
  const dateFrom = new Intl.DateTimeFormat("sv-SE", {
    timeZone: "Asia/Tokyo",
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  }).format(start);
  return { dateFrom, dateTo };
}
